client/components/MobileAware/MobileAware.tsx

```diff
@@ -0,0 +1,61 @@
+import React, { useState, useEffect } from 'react';
+
+import { useViewport } from 'client/utils/useViewport';
+
+require('./mobileAware.scss');
+
+export type RenderProps = { ref?: React.Ref<any>; className: string };
+type Renderable = null | React.ReactElement | ((props: RenderProps) => React.ReactElement);
+
+type Props = {
+	desktop?: Renderable;
+	mobile?: Renderable;
+};
+
+const componentPrefix = 'mobile-aware-component';
+
+const renderMobileOrDesktop = (
+	renderable: Renderable,
+	ref: React.Ref<unknown>,
+	isFirstMount: boolean,
+	isMobileViewport: null | boolean,
+	isMobileVersion: boolean,
+) => {
+	if (!renderable) {
+		return null;
+	}
+	const matchesViewportSize = isMobileViewport === isMobileVersion;
+	// During server-side rendering, isFirstMount is true, so we unconditionally render
+	// both the provided mobile and the desktop versions, and rely on CSS to show only the
+	// correct one. During the first client render, we already know what viewport size we have,
+	// but we still need to render both elements because React insists that the first client
+	// render match the server render exactly. After that, we can stop rendering one of them.
+	if (!isFirstMount && !matchesViewportSize) {
+		return null;
+	}
+	const renderProps: RenderProps = {
+		className: isMobileVersion ? `${componentPrefix}__mobile` : `${componentPrefix}__desktop`,
+		...(matchesViewportSize ? { ref } : null),
+	};
+	if (typeof renderable === 'function') {
+		return renderable(renderProps);
+	}
+	return React.cloneElement(renderable, renderProps);
```

Comment:
why not `return <renderable {...renderProps} />;`?

Reply:
This is `React.cloneElement()` not `React.createElement()`!

Reply:
I see your `cloneElement` and I raise you (the question) `return <renderable.type {...renderProps} />;`

per [react top-level api documentation](https://reactjs.org/docs/react-api.html#cloneelement) these things should be nearly equivalent. The difference is the [ben]-ne-sais-quoi of preserving refs.

Reply:
It never occurred to me that you could do `<node.type .../>`! `React.cloneElement(node)` is the way I learned how to do this. I'm kind of getting lost in their explanation around refs there, but we do definitely need React to pass along the `ref` from `renderProps` in this case.

Editorializing: I think that example is just there to build intuition and that you should treat React elements (the return value of `React.createElement(X)` or `<X/>`) as black boxes, and avoid inspecting their properties (e.g. `type`) at all costs! I don't think those are part of the public React API.

Reply:
My two cents: disregarding refs, I feel like `cloneElement` is a bit more elegant/easier on the eyes than

```tsx
<renderable.type {...renderable.props} {...renderProps} />
```